lessons/advanced/typespec.md

```diff
@@ -0,0 +1,127 @@
+---
+layout: page
+title: Specifications and types
+category: advanced
+order: 9
+lang: en
+---
+
+In this lesson we will learn about `@spec` and `@type` syntax. First is more syntax complement for writing documentation that could by analyzed by tools. Second helps us to write more readable and easier to understand code.
+
+{% include toc.html %}
+
+## Introduction 
+
+It's not uncommon you would like to describe interface of your function. Of course You can use [@doc annotation](/lessons/basic/documentation), but it is only information for other developers that is not checked in compilation time. For this purpose Elixir has `@spec` annotation to describe specification of function that will be checked by compiler.
+
+However in some cases specification is going to be quite big and complicated. If you would like to reduce complexity, you want to introduce custom type definition. Elixir has `@type` annotation for that. In the other hand, Elixir is still dynamic language. That means all information about type will be ignored by compiler, but could be used by other tools.   
+
+## Specification
+
+If you have experience with Java or Ruby you could think about specification as an `interface`. Specification defines what should be type of function parameters and return value.
```

Comment:
I know were straying away from the comparisons to Ruby, and that this also throws in Java's hat, but should we remove it and just explain the concept of an interface agnostically if its useful as an analogy?